Sources/Container.MainActor.erb

```diff
@@ -0,0 +1,82 @@
+//
+//  Copyright © 2024 Swinject Contributors. All rights reserved.
+//
+
+//
+// NOTICE:
+//
+// Container.MainActor.swift is generated from Container.MainActor.erb by ERB.
+// Do NOT modify Container.MainActor.swift directly.
+// Instead, modify Container.MainActor.erb and run `script/gencode` at the project root directory to generate the code.
+//
+
+<% arg_count = 9 %>
+import Foundation
+
+// MARK: - MainActor registration
+@available(iOS 13.0, macOS 10.15, *)
+extension Container {
+
+    /// Adds a registration for the specified service with the factory closure to specify how the service is
+    /// resolved with dependencies which must be resolved on the main actor.
+    ///
+    /// - Parameters:
+    ///   - serviceType:        The service type to register.
+    ///   - name:               A registration name, which is used to differentiate from other registrations
+    ///                         that have the same service and factory types.
+    ///   - mainActorFactory:   The @MainActor closure to specify how the service type is resolved with the dependencies of the type.
+    ///                         It is invoked when the ``Container`` needs to instantiate the instance.
+    ///                         It takes a ``Resolver`` to inject dependencies to the instance,
+    ///                         and returns the instance of the component type for the service.
+    ///
+    /// - Returns: A registered ``ServiceEntry`` to configure more settings with method chaining.
+    @discardableResult
+    public func register<Service>(
+        _ serviceType: Service.Type,
+        name: String? = nil,
+        mainActorFactory: @escaping @MainActor (Resolver) -> Service
+    ) -> ServiceEntry<Service> {
+        return register(serviceType, name: name) { r in
+            MainActor.assumeIsolated {
+                return mainActorFactory(r)
+            }
```

Comment:
Only calling `MainActor.assumeIsolated` will crash when called _off_ of the main thread.

To avoid a crash, you could check to see if execution is already on the main thread, and then hop to it if needed:
```suggestion
            onMainActor {
               mainActorFactory(r)
            }
```

Where onMainActor is defined as:
```swift
func onMainActor<ReturnValue>(work: () -> ReturnValue) -> ReturnValue {
    if Thread.isMainThread {
        return MainActor.assumeIsolated { work() }
    }
    return DispatchQueue.main.sync {
        MainActor.assumeIsolated { work() }
    }
}
```